### PR TITLE
[WIP] Add ModernBERT implementation in examples

### PR DESCRIPTION
**Pull request by vctrmn**
# Description

This is a work-in-progress implementation of ModernBERT. ModernBERT is an optimized variant of BERT that includes:
- GeGLU activation function instead of GeLU
- Improved attention mechanism with local and global attention patterns
- RoPE (Rotary Position Embeddings) for better position encoding
- ...

References:
https://huggingface.co/blog/modernbert
https://huggingface.co/docs/transformers/main/en/model_doc/modernbert
https://github.com/AnswerDotAI/ModernBERT
https://github.com/huggingface/transformers/blob/main/src/transformers/models/modernbert/modular_modernbert.py

## TODO

- [x] Embeddings layer with layer normalization
- [x] Attention mechanisms (both local and global variants)
- [x] MLP with GeGLU activation
- [x] Basic test implementation
- [ ] Create ModernBertOptions struct to replace hardcoded constants:
- [ ] Implement Flash Attention 2 optimization for better throughput
- [ ] Create geglu function in tensor.zig
- [ ] Handle layer 0 special case using a union type for both "void" and zml.nn.LayerNorm cases
- [ ] Implement proper position_ids handling (currently using null)
- [ ] Handle global and local attention properly in ModernBertEncoderLayer
- [ ] Complete layers field in ModernBertModel (currently commented out)
- [ ] ... 

## Testing

Basic testing has been implemented for:
- Embeddings layer
- Attention mechanisms (both local and global)
- MLP layer
- Layer normalization
- Full encoder layer (with only local attention mechanism + without handling layer 0)

Current test tolerances range from 1e-6 to 2e-3 depending on the layer.

```bazel run -c opt //modernbert:test-implementation```

__Feedback and suggestions are welcome !__

### CODE REVIEW

File: examples/modernbert/modernbert.zig [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    pub fn forward(self: ModernBertEmbeddings, input_ids: Tensor) Tensor {
+        // Perform tok_embeddings
+        const hidden_states = zml.call(self.tok_embeddings, .forward, .{input_ids});
+
+        // Perform norm
+        return zml.call(self.norm, .forward, .{hidden_states});
+    }
+};
+
+// TODO: Create geglu fn in tensor.zig ?
```
Comment (gwenzek) [inline examples/modernbert/modernbert.zig:33]: the right place to add this would be `nn.zig`

File: examples/modernbert/modernbert.zig [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        // Project to query, key, value - { batch_size, seq_len, 3 * num_heads * head_dim }
+        var qkv: Tensor = zml.call(self.Wqkv, .forward, .{hidden_states}); // Wqkv.out.0
+
+        // Reshape to { batch_size, seq_len, 3, num_heads, head_dim }
+        qkv = qkv.reshape(.{ batch_size, seq_length, 3, num_heads, head_dim });
+        // Transpose for attention computation  { batch_size, seq_len, 3, num_heads, head_dim } -> { batch_size, num_heads, 3, seq_length, head_dim }
+        qkv = qkv.transpose(.{ 0, 3, 2, 1, 4 });
+
+        // Split into query, key, value tensors - each { batch_size, num_heads, seq_length, head_dim }
+        var q = qkv.slice1d(2, .{ .start = 0, .end = 1 }).squeeze(2).withTags(.{ .b, .h, .s, .hd });
```
Comment (gwenzek) [inline examples/modernbert/modernbert.zig:98]: I'd use `var q, var k, var v = qkv.chunkExact(qkv, 2)`

File: examples/modernbert/modernbert.zig [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+        // Apply rotary position embeddings (RoPE)
+        // Layer 0, 3, 6, 9, 12 ... use global RoPE
+        // Layer 1, 2, 4, 5, 7, 8, 10, 11 ... use local RoPE
+        const rope_opts = zml.nn.RopeOpts{
+            .impl = .sequential,
+            .freq_base = if (self.is_global_attention) 160_000 else 10_000,
+        };
+        const pos_idx = null; // TODO: This is temporary, need to implement position_ids ?
+        _ = position_ids;
```
Comment (gwenzek) [inline examples/modernbert/modernbert.zig:110]: you don't need to compute position id. Position id allow to compute a slice of the Rope matrix. But in a text encoder you can generate the full matrix at once, which correspond to `pos_idx = null`

I need to update the rope documenation to the new api

File: examples/modernbert/modernbert.zig [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        var qkv: Tensor = zml.call(self.Wqkv, .forward, .{hidden_states}); // Wqkv.out.0
+
+        // Reshape to { batch_size, seq_len, 3, num_heads, head_dim }
+        qkv = qkv.reshape(.{ batch_size, seq_length, 3, num_heads, head_dim });
+        // Transpose for attention computation  { batch_size, seq_len, 3, num_heads, head_dim } -> { batch_size, num_heads, 3, seq_length, head_dim }
+        qkv = qkv.transpose(.{ 0, 3, 2, 1, 4 });
+
+        // Split into query, key, value tensors - each { batch_size, num_heads, seq_length, head_dim }
+        var q = qkv.slice1d(2, .{ .start = 0, .end = 1 }).squeeze(2).withTags(.{ .b, .h, .s, .hd });
+        var k = qkv.slice1d(2, .{ .start = 1, .end = 2 }).squeeze(2).withTags(.{ .b, .h, .s, .hd });
```
Comment (gwenzek) [inline examples/modernbert/modernbert.zig:99]: i'd recommend to tag earlier in your function so you don't need the transpose(.{ 0, 3, 2, 1, 4 })

**Comment by gwenzek** [comment]
Nice work ! I left a few tips, but don't hesitate to asks more specific questions.

your precision sounds good for now, but you'll probably have more issues for the attn.

wrt flash attention don't worry too much about this because the CUDA backend detects attention and uses relevant cudnn kernels.
Once you've a working implementation, I'd be happy to pass it through the profiler to check that the code is correctly compiled.

**Comment by vctrmn** [comment]
Thanks @gwenzek for your feedback ! I also need guidance on the best way to handle layer0 case and global vs local attention alternance patterns. 
- Layer 0: no attention norm
- Global/local attention alternation (global on 0,3,6,9...)

I tried using ModernBertModel's init fn but the logs from testLayer exec show that it is not being called (I don't see the layer initialization logs) :
```zig
pub const ModernBertModel = struct {
    embeddings: ModernBertEmbeddings,
    layers: []ModernBertEncoderLayer,
    final_norm: zml.nn.LayerNorm,

    pub fn init(self: *ModernBertModel) void {
        for (self.layers, 0..) |*encoder_layer, layer_idx| {
            log.info("layer_idx : {}", .{layer_idx});
            if (layer_idx == 0) encoder_layer.attn_norm = void;
            encoder_layer.attn.is_global_attention = (layer_idx % 3 == 0);
        }
    }

    pub fn forward(self: ModernBertModel, input_ids: Tensor, attention_mask: Tensor) Tensor {
        _ = attention_mask; // autofix

        for (self.layers, 0..) |*encoder_layer, layer_idx| {
            log.info("encoder_layer {} attn_norm {} and is_global {}", .{ layer_idx, encoder_layer.attn_norm, encoder_layer.attn.is_global_attention });
        }

        return Tensor.constant(input_ids.shape(), input_ids.dtype().zero());
    }
};

```

Any idea on how to handle this ?

**Comment by gwenzek** [comment]
> I tried using ModernBertModel's init fn but the logs from testLayer exec show that it is not being called (I don't see the layer initialization logs) :

You can always call `init` yourself on the ModernBertModel struct returned by `populateModelWithPrefix`
That's how llama.zig  does it:

https://github.com/zml/zml/blob/cf135bd4281c75110fa33501ac87f216c7c3edb2/examples/llama/main.zig#L178-L202

File: examples/modernbert/modernbert.zig [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    const target_shape = zml.Shape.init(.{ .tgt = tgt_seq_len, .src = src_seq_len }, global_attention_mask.dtype());
+
+    // TODO: Confirm this is the correct way to do this : reshape (or insertAxes, appendAxes) instead of unsqueeze ?
+
+    // Create position indices (for rows and cols)
+    // Original py code : rows = torch.arange(global_attention_mask.shape[2]).unsqueeze(0)
+    var rows = Tensor.arange(.{ .end = tgt_seq_len }, global_attention_mask.dtype())
+        .unsqueeze(0);
+    // var rows = Tensor.arange(.{ .end = tgt_seq_len }, global_attention_mask.dtype())
+    //     .reshape(.{ 1, tgt_seq_len });
```
Comment (vctrmn) [inline examples/modernbert/modernbert.zig:184]: I have the impression that the `modern_bert_model` test runs faster using reshape

**Comment by vctrmn** [comment]
With `1e-2` tolerance

```
Testing model layer:
info(zml/testing): Testing model.model
info(zml): Compiling modernbert.ModernBertModel.forward with { Shape({1,9,i64}), Shape({1,9,i64}) }
info: global_attention_mask : Tensor({b=1,broadcast_dim=1,tgt=9,src=9,f32})
info: rows : Tensor({1,9,f32})
info: cols : Tensor({9,1,f32})
info: distance : Tensor({tgt=9,src=9,f32})
info: sliding_window_mask : Tensor({b=1,broadcast_dim=1,tgt=9,src=9,f32})
info: Global attention
info: attn_output : Tensor({b=1,s=9,d=768,f32})
info: Local attention
info: attn_output : Tensor({b=1,s=9,d=768,f32})
info: attn_output : Tensor({b=1,s=9,d=768,f32})
info(zml/module): Compilation took 1.055s
info(zml/testing): prefix: model.model.in.0
info(zml/testing): prefix: model.model.in.1
info(zml/testing): all good for model.model !
```

File: examples/modernbert/modernbert.zig [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+        return output;
+    }
+};
+
+pub fn generateSlidingWindowMask(global_attention_mask: Tensor) Tensor {
+    const tgt_seq_len = global_attention_mask.dim(.tgt);
+    const src_seq_len = global_attention_mask.dim(.src);
+    const target_shape = zml.Shape.init(.{ .tgt = tgt_seq_len, .src = src_seq_len }, global_attention_mask.dtype());
+
+    // TODO: Confirm this is the correct way to do this : reshape (or insertAxes, appendAxes) instead of unsqueeze ?
```
Comment (vctrmn) [inline examples/modernbert/modernbert.zig:177]: @gwenzek Let me know the correct way when you have a moment 🤔

Comment (gwenzek) [reply examples/modernbert/modernbert.zig:157]: sorry I missed that, my GH notifications are a bit of a mess.

I think here I'd use `iota`: 

```
const mask_shape = Shape.init(.{ .tgt = tgt_seq_len, .src = src_seq_len });
const rows = Tensor.iota(mask_shape, .q);
const cols = Tensor.iota(mask_shape, .k);
```

but you're code isn't bad, just a bit "pytorch like" to my taste.

I'm not sure what all the axes correspond to so it's hard for me to give names. 
My brain is also a bit toast atm, I should have better ideas on Monday morning :-)

**Comment by vctrmn** [comment]
I will probably need some help on this one ...

```logs
info(modernbert): ✅	Loaded weights in 1227ms
info(modernbert): ✅	Compiled model in 4230ms
info(modernbert): 	Input text: Zig is the [MASK] programming language.
info(modernbert): Dimensions:
info(modernbert): 	batch_size: 1, seq_len: 64, vocab_size: 50396
info(modernbert): 	mask_position: 8
info(modernbert): 	base_offset: 403168 = 8 * 50396
info(modernbert): 	output_buffer size: 3223552
info(modernbert): First logits:
info(modernbert): 	logit[0] = 3.753852 (token: '|||IP_ADDRESS|||')
info(modernbert): 	logit[1] = 4.349829 (token: '<|padding|>')
info(modernbert): 	logit[2] = 3.610769 (token: '!')
info(modernbert): 	logit[3] = 6.032334 (token: '"')
info(modernbert): 	logit[4] = 4.437929 (token: '#')
info(modernbert): Top predictions:
info(modernbert): 	score: 0.851775, token: '                       '
error: OutOfMemory
```

**Comment by vctrmn** [comment]
@gwenzek 

PR ready for review! Two small tasks remain:

- [] CCompile with max_seq_length and pad shorter input text with [PAD] token
- [] Test multi-platform activations file generation

**Comment by vctrmn** [comment]
Some interesting logs : 
On my mac Intel : `bazel run -c opt //modernbert:ModernBERT-base`
```logs
info(zml/context): Available Platforms:
info(zml/context):   ✅ cpu (AUTO-SELECTED)
info(zml/context):        ◦ #0: cpu
info(zml/context):   •  cuda
info(zml/context):   •  rocm
info(zml/context):   •  tpu
info(zml/context):   •  neuron
info(modernbert):       ModernBERT options: modernbert.ModernBertOptions{ .num_attention_heads = 12, .tie_word_embeddings = true }
info(modernbert):       Loading tokenizer from ../zml++huggingface+ModernBERT-base/tokenizer.json
info(modernbert): ✅    Loaded tokenizer from ../zml++huggingface+ModernBERT-base/tokenizer.json [103.033ms]
info(modernbert):       Loading ModernBERT weights from ../zml++huggingface+ModernBERT-base/model.safetensors...
info(modernbert): ✅    Loaded weights in 2809ms
info(modernbert):       Compiling ModernBERT model...
2025-02-19 14:26:08.406986: I xla/service/dump.cc:569] HloModule dump enabled with path prefix: , suffix: before_optimizations
info(modernbert): ✅    Loaded weights and compiled model in 6381ms
info(modernbert):       Input text: Paris is the [MASK] of France.
info(modernbert): ⏱️    Model inference in  2359ms
info(modernbert): ✅    Top 5 predictions:
info(modernbert):         • score: 0.9891  word: ' capital' token: 5347
info(modernbert):         • score: 0.0060  word: ' Capital' token: 17572
info(modernbert):         • score: 0.0022  word: ' city' token: 2846
info(modernbert):         • score: 0.0005  word: ' center' token: 4055
info(modernbert):         • score: 0.0003  word: ' symbol' token: 9484
```
On my VM (with L4 GPU) : `bazel run -c opt --@zml//runtimes:cuda=true //modernbert:ModernBERT-base`
```logs
info(zml/context): Available Platforms:
info(zml/context):   •  cpu
info(zml/context):   ✅ cuda (AUTO-SELECTED)
info(zml/context):        ◦ #0: NVIDIA L4
info(zml/context):   •  rocm
info(zml/context):   •  tpu
info(zml/context):   •  neuron
info(modernbert):       ModernBERT options: modernbert.ModernBertOptions{ .num_attention_heads = 12, .tie_word_embeddings = true }
info(modernbert):       Loading tokenizer from ../zml++huggingface+ModernBERT-base/tokenizer.json
info(modernbert): ✅    Loaded tokenizer from ../zml++huggingface+ModernBERT-base/tokenizer.json [66.504ms]
info(modernbert):       Loading ModernBERT weights from ../zml++huggingface+ModernBERT-base/model.safetensors...
info(modernbert): ✅    Loaded weights in 628ms
info(modernbert):       Compiling ModernBERT model...
info(modernbert): ✅    Loaded weights and compiled model in 904ms
info(modernbert):       Input text: Paris is the [MASK] of France.
info(modernbert): ⏱️    Model inference in  68ms
info(modernbert): ✅    Top 5 predictions:
info(modernbert):         • score: 0.9891  word: ' capital' token: 5347
info(modernbert):         • score: 0.0060  word: ' Capital' token: 17572
info(modernbert):         • score: 0.0022  word: ' city' token: 2846
info(modernbert):         • score: 0.0005  word: ' center' token: 4055
info(modernbert):         • score: 0.0003  word: ' symbol' token: 9484
```

**Comment by gwenzek** [comment]
Sorry I missed the GH notifications, I thought you were still working on it.
It looks pretty great. I'm gonna merge this. Thanks a lot :-)

**Review by gwenzek — APPROVED** [review]

**Comment by gwenzek** [comment]
So, I wanted to refactor a bit the mask generation logic and found a bug in your implem. Your local attention mask where always the same than the global mask.

But when fixing that it surfaced a bug in zml sdpa, where if you mask a full row you get Nan, which then propagate to every where.

I've added a fix in zml softmax for that.

**Comment by gwenzek** [comment]
Thanks a lot, this has been merged in #192
